core/src/Kokkos_Parallel.hpp

```diff
@@ -71,6 +72,35 @@
 namespace Kokkos {
 namespace Impl {
 
+template <class T, class = void>
+struct is_detected_execution_space : std::false_type {
+  using type = not_a_type;
+};
+
+template <class T>
+struct is_detected_execution_space<T, void_t<typename T::execution_space>>
+    : std::true_type {
+  using type = typename T::execution_space;
+};
+
+template <class T>
+using detected_execution_space_t =
+    typename is_detected_execution_space<T>::type;
+
+template <class T, class = void>
+struct is_detected_device_type : std::false_type {
+  using type = not_a_type;
+};
+
+template <class T>
+struct is_detected_device_type<T, void_t<typename T::device_type>>
+    : std::true_type {
+  using type = typename T::device_type;
+};
+
+template <class T>
+using detected_device_type_t = typename is_detected_device_type<T>::type;
+
```

Comment:
Why not just put `is_detected` in directly instead of making special versions of each of these?